### PR TITLE
Insert many

### PR DESCRIPTION
**Pull request by Mattchewone**
### Summary

Change the way the insertion of many items using mongoose works to use the mongdb native driver and the insertMany method. 

This currently uses the `feathers-batch` response style of an array, the first array item being an array of errors or null, and the second array item being an array of successful items or null.

### CODE REVIEW

**Comment by daffl** [comment]
Thank you for the pull request @Mattchewone a couple of questions:

- Could you elaborate the reasons for this change?
- Could this be done in a hook instead of changing the core adapter?
- The service tests are failing and it seems like it is breaking the behaviour expected of a database adapter

**Comment by Mattchewone** [comment]
Hey @daffl I am not sure why all of the tests failed on Travis, locally mine only failed on 2. 

The reason for the change is that mongoose for batch create will run a save on each array item, where as switching to mongodb native insertMany will send all to the database in one request rather than many. I thought this was a better performance. 

Utilising this also means that we can see which of the items within the batch failed during the request, which is why I altered the return array. I thought I would follow the feathers-batch pattern in returning an array of errors first, then an array of successful inserts. 

I wasn't too sure how else to return the errors that happened within the insert. 

For the expected behaviour of a database adapter, what is expected of a batch insert, how is the user/service to know which of the items within the batch have successfully inserted and which have failed. 
Should the whole request fail if any of them fail?

The issue I was having was with a batch of 500 items, if 2 of them failed the return would be only one of the errors and this would then skip any after hooks because of the errors raised and would go straight to error hooks. This meant that I was unable to run any after hooks on any of the successful inserts.

**Comment by marshallswain** [comment]
I think it's a good idea to have this baked into the adapter.  It's kinda problematic, though, because of what he mentioned: how do we handle bulk insert errors?

**Comment by marshallswain** [comment]
We've updated the PR to use the existing standard return format for successful inserts.  It pushes any failed inserts into `params.errors`.  The param used is configurable using `options.bulkErrorsKey`.  So setting `params.bulkErrorsKey = 'specialErrors'` will cause errors to be stored at `params.specialErrors`.

Each error will have the format:

```js
{ 
  type: 'ValidationError' || 'WriteError',
  data: {/*the original record*/}
}
```

File: src/service.js [hunk 1]
```diff
@@ -184,6 +190,168 @@ class Service {
...
+        // Add the errors results to params.errors
+        params[this.bulkErrorsKey] = params[this.bulkErrorsKey].concat(errors);
+
+        return data;
+      })
+      .catch(errorHandler);
+  }
+
+  _insertMany (data) {
+    const INSERT_MANY_CONVERT_OPTIONS = {
```
Comment (ekryski) [inline src/service.js]: What if someone needs to change these? They aren't exposed anywhere.

Comment (corymsmith) [reply src/service.js]: Agreed, should potentially be an option that gets passsed in.

**Comment by ekryski** [comment]
This is a pretty big change, with some duplicate logic in here. If we are going to add better support for inserting multiple objects can we see some benchmarks? I'd like to see how much we are actually gaining in perf when adding this much complexity.

Currently you can already accomplish what you want in terms of not failing on first error insert using `feathers-batch` and this DB adapter as is. Admittedly that will likely be slower but I'd like to know how much slower and how often the use case for massive bulk insert appears.

In a lot of real world apps we haven't seen too many instances where you are doing bulk creates beyond 5-10 records except for 4 cases:

1. Seed data/DB migrations.
2. Background workers that are inserting/migrating/updating data similar to seeding
3. Processing bulk file uploads (ie.CSV, XML, etc.)
4. Clients regaining connectivity and sending a batch of offline interactions when they regain connectivity

In the first 2 cases, I would argue ms level performance differences are not a huge deal. In the 3rd and 4th cases, this is reasonable but likely not as common compared to "normal" CRUD, and you can already solve this 3 different ways:

1. drop down to the native MongoDB driver
2. call the `hook.service.Model.insertMany` in a `before` hook and set `hook.result` with your data and bypass the service call; or
3. [extend the service](https://docs.feathersjs.com/api/databases/common.html#extending-adapters).

Personally, I'd like to see some numbers first. Otherwise this should likely be implemented as one of the above options instead of doubling the adapter size for what I would consider is a 20% or less use case. Likely option 2 is best.

Happy to be proven wrong, but in that case I think we can come up with a better means of processing bulk actions across all adapters.

File: src/service.js (hunk 1 above)
Comment (marshallswain) [reply src/service.js]: We can expose these when the need arises.

**Comment by marshallswain** [comment]
I’ll just revive this as `feathers-mongoose-advanced` for now.  If it turns out to be a successful experiment, we can bring over its features like we did with the query syntax and other goodies from the first iteration of `feathers-mongoose-advanced`.

**Comment by ekryski** [comment]
@marshallswain any chance you can just accomplish this in a hook? Would be a much easier integration and we can write tests for the hook itself. Unless of course you want to emit events for every item in a batch creation. Not sure how common that is, or if it's part of your use case.

**Pull request closed without merging**